### PR TITLE
Replay

### PR DESCRIPTION
**Pull request by KirillOsenkov**
Binary log replay functionality. 

Instead of specifying a project or solution, you can specify a .binlog file to "build" and it will replay all the events in the source log into the usual loggers.

Note that BuildLogReplayEventArgs and its base class EventArgsDispatcher need to be made public for the replay functionality in MSBuild.exe to work. I've made the class sealed and added <remarks> with explanations.

### CODE REVIEW

**Comment by KirillOsenkov** [comment]
@dotnet-bot test Windows_NT Build for Desktop please
@dotnet-bot test Windows_NT Build for CoreCLR please

**Comment by KirillOsenkov** [comment]
@dotnet-bot test Ubuntu16.04 Build for CoreCLR

File: src/Build/Logging/BinaryLogger/BinaryLogReplayEventSource.cs [hunk 1]
```diff
@@ -10,65 +10,54 @@ namespace Microsoft.Build.Logging
...
-                    Dispatch(instance);
+                        Dispatch(instance);
+                    }
                 }
             }
+            catch (Exception ex)
+            {
+                string code;
+                string helpKeyword;
+                var text = ResourceUtilities.FormatResourceString(out code, out helpKeyword, "InvalidLogFileFormat", ex.Message);
```
Comment (cdmihai) [inline src/Build/Logging/BinaryLogger/BinaryLogReplayEventSource.cs]: there appears to be a FormatResourceString overload that does not have the out params

File: src/MSBuild/XMake.cs [hunk 2]
```diff
@@ -598,26 +598,36 @@ string commandLine
...
                     DateTime t1 = DateTime.Now;
 
+                    // If the primary file passed to MSBuild is a .binlog file, play it back into passed loggers
+                    // as if a build is happening
+                    if (FileUtilities.IsBinaryLogFilename(projectFile))
+                    {
+                        ReplayBinaryLog(projectFile, loggers);
+                    }
+                    else // regular build
+                    {
```
Comment (cdmihai) [inline src/MSBuild/XMake.cs]: The new else block is not indented.

**Comment by KirillOsenkov** [comment]
There's already a test for replay here:
https://github.com/Microsoft/msbuild/blob/xplat/src/Build.UnitTests/BinaryLogger_Tests.cs#L46

**Pull request merged**